### PR TITLE
[BEAM-2405] Write to BQ using the streaming API

### PR DESCRIPTION
**Pull request by sb2nov**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [ ] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [ ] Make sure tests pass via `mvn clean verify`.
 - [ ] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

This should be ready for review so that we can test it with streaming pipelines

There are few followup items after this PR:
- Integration with the Dataflow runner to work in both Batch and Streaming
- Migrate the tornadoes example to use this and deprecate the sink interface

### CODE REVIEW

**Comment by sb2nov** [comment]
R: @chamikaramj PTAL

File: sdks/python/apache_beam/io/gcp/bigquery.py [hunk 1]
```diff
@@ -813,6 +817,7 @@ def _create_table(self, project_id, dataset_id, table_id, schema):
     request = bigquery.BigqueryTablesInsertRequest(
         projectId=project_id, datasetId=dataset_id, table=table)
     response = self.client.tables.Insert(request)
+    logging.info("Created the table with id %s", table_id)
```
Comment (chamikaramj) [inline sdks/python/apache_beam/io/gcp/bigquery.py]: logging.debug ?

"Created a table"

File: sdks/python/apache_beam/io/gcp/bigquery.py [hunk 2]
```diff
@@ -1134,3 +1139,110 @@ def convert_row_to_dict(self, row, schema):
...
     return result
+
+
+class BigQueryWriteFn(DoFn):
+  """A ``DoFn`` that streams writes to BigQuery once the table is created.
+  """
+
+  def __init__(self, table_id, dataset_id, project_id, batch_size, schema,
+               create_disposition, write_disposition, client):
+    self.table_id = table_id
```
Comment (chamikaramj) [inline sdks/python/apache_beam/io/gcp/bigquery.py]: Please add a doc comment.

File: sdks/python/apache_beam/io/gcp/bigquery.py [hunk 3]
```diff
@@ -1134,3 +1139,110 @@ def convert_row_to_dict(self, row, schema):
...
+    self.create_disposition = create_disposition
+    self.write_disposition = write_disposition
+    self._rows_buffer = []
+    self._max_batch_size = batch_size or 500
+
+  def start_bundle(self):
+    self._rows_buffer = []
+    # Transform the table schema into a bigquery.TableSchema instance.
+    if isinstance(self.schema, basestring):
+      # TODO(silviuc): Should add a regex-based validation of the format.
```
Comment (chamikaramj) [inline sdks/python/apache_beam/io/gcp/bigquery.py]: Are we still hoping to do this (TODO) ?

File: sdks/python/apache_beam/io/gcp/bigquery.py [hunk 4]
```diff
@@ -1134,3 +1139,110 @@ def convert_row_to_dict(self, row, schema):
...
+      table_schema = bigquery.TableSchema()
+      schema_list = [s.strip(' ') for s in self.schema.split(',')]
```
Comment (chamikaramj) [inline sdks/python/apache_beam/io/gcp/bigquery.py]: s.strip()

File: sdks/python/apache_beam/io/gcp/bigquery.py [hunk 5]
```diff
@@ -1134,3 +1139,110 @@ def convert_row_to_dict(self, row, schema):
...
+      for field_and_type in schema_list:
+        field_name, field_type = field_and_type.split(':')
+        field_schema = bigquery.TableFieldSchema()
+        field_schema.name = field_name
+        field_schema.type = field_type
+        field_schema.mode = 'NULLABLE'
```
Comment (chamikaramj) [inline sdks/python/apache_beam/io/gcp/bigquery.py]: Do we support other modes ?

File: sdks/python/apache_beam/io/gcp/bigquery.py (hunk 4 above)
Comment (chamikaramj) [inline sdks/python/apache_beam/io/gcp/bigquery.py]: Is this form of schema sufficient or should we try to support a more advanced form (can be a separate PR) ?

File: sdks/python/apache_beam/io/gcp/bigquery.py [hunk 6]
```diff
@@ -1134,3 +1139,110 @@ def convert_row_to_dict(self, row, schema):
...
+        table_schema.fields.append(field_schema)
+      self.table_schema = table_schema
+    elif self.schema is None:
+      # TODO(silviuc): Should check that table exists if no schema specified.
```
Comment (chamikaramj) [inline sdks/python/apache_beam/io/gcp/bigquery.py]: Please implement or remove this TODO.

File: sdks/python/apache_beam/io/gcp/bigquery.py [hunk 7]
```diff
@@ -1134,3 +1139,110 @@ def convert_row_to_dict(self, row, schema):
...
+      self.table_schema = self.schema
+    elif isinstance(schema, bigquery.TableSchema):
+      self.table_schema = self.schema
+    else:
+      raise TypeError('Unexpected schema argument: %s.' % self.schema)
+
+    self.bigquery_wrapper = BigQueryWrapper(client=self.client)
+    self.bigquery_wrapper.get_or_create_table(
```
Comment (chamikaramj) [inline sdks/python/apache_beam/io/gcp/bigquery.py]: Does BQ guarantee consistency here ? Please add a comment.

File: sdks/python/apache_beam/io/gcp/bigquery.py [hunk 8]
```diff
@@ -1134,3 +1139,110 @@ def convert_row_to_dict(self, row, schema):
...
+    # Flush the current batch of rows to BigQuery.
+    passed, errors = self.bigquery_wrapper.insert_rows(
+        project_id=self.project_id, dataset_id=self.dataset_id,
+        table_id=self.table_id, rows=self._rows_buffer)
+    if not passed:
+      raise RuntimeError('Could not successfully insert rows to BigQuery'
+                         ' table [%s:%s.%s]. Errors: %s'%
+                         (self.project_id, self.dataset_id,
+                          self.table_id, errors))
+    logging.info("Successfully wrote %d rows.", len(self._rows_buffer))
```
Comment (chamikaramj) [inline sdks/python/apache_beam/io/gcp/bigquery.py]: Let's remote this log for successful writes or make it logging.debug.

File: sdks/python/apache_beam/io/gcp/bigquery.py [hunk 9]
```diff
@@ -1134,3 +1139,110 @@ def convert_row_to_dict(self, row, schema):
...
+    self._rows_buffer = []
+
+
+class WriteToBigQuery(PTransform):
+
+  def __init__(self, table, dataset=None, project=None, schema=None,
+               create_disposition=BigQueryDisposition.CREATE_IF_NEEDED,
```
Comment (chamikaramj) [inline sdks/python/apache_beam/io/gcp/bigquery.py]: Please add a doc comment.

File: sdks/python/apache_beam/io/gcp/bigquery.py [hunk 10]
```diff
@@ -1134,3 +1139,110 @@ def convert_row_to_dict(self, row, schema):
...
+               create_disposition, write_disposition, client):
+    self.table_id = table_id
+    self.dataset_id = dataset_id
+    self.project_id = project_id
+    self.schema = schema
+    self.client = client
+    self.create_disposition = create_disposition
+    self.write_disposition = write_disposition
+    self._rows_buffer = []
+    self._max_batch_size = batch_size or 500
```
Comment (chamikaramj) [inline sdks/python/apache_beam/io/gcp/bigquery.py]: Why did you choose 500 for default batch size ? Please add a comment. (should we run some experiments to see if changing this has any affect ? )

File: sdks/python/apache_beam/io/gcp/bigquery.py [hunk 11]
```diff
@@ -1134,3 +1139,110 @@ def convert_row_to_dict(self, row, schema):
...
+    bigquery_write_fn = BigQueryWriteFn(
+        table_id=self.table_reference.tableId,
+        dataset_id=self.table_reference.datasetId,
+        project_id=self.table_reference.projectId,
+        batch_size=self.batch_size,
+        schema=self.schema,
+        create_disposition=self.create_disposition,
+        write_disposition=self.write_disposition,
+        client=self.test_client)
+    return pcoll | 'Write to BQ' >> ParDo(bigquery_write_fn)
```
Comment (chamikaramj) [inline sdks/python/apache_beam/io/gcp/bigquery.py]: BigQuery instead of BQ here ?

File: sdks/python/apache_beam/io/gcp/bigquery_test.py [hunk 12]
```diff
@@ -824,6 +824,138 @@ def test_get_or_create_dataset_fetched(self):
...
+    fn.start_bundle()
+    self.assertTrue(client.tables.Get.called)
+    self.assertTrue(client.tables.Insert.called)
+
+  def test_dofn_client_process_flush_not_called(self):
+    client = mock.Mock()
+    client.tables.Get.return_value = bigquery.Table(
+        tableReference=bigquery.TableReference(
+            projectId='project_id', datasetId='dataset_id', tableId='table_id'))
+    client.tabledata.InsertAll.return_value = \
```
Comment (chamikaramj) [inline sdks/python/apache_beam/io/gcp/bigquery_test.py]: I believe we usually use () instead of \ for line breaks.

File: sdks/python/apache_beam/io/gcp/bigquery_test.py [hunk 13]
```diff
@@ -824,6 +824,138 @@ def test_get_or_create_dataset_fetched(self):
...
+        schema='month:INTEGER',
+        create_disposition=create_disposition,
+        write_disposition=write_disposition,
+        client=client)
+
+    fn.start_bundle()
+    self.assertTrue(client.tables.Get.called)
+    self.assertTrue(client.tables.Insert.called)
+
+  def test_dofn_client_process_flush_not_called(self):
```
Comment (chamikaramj) [inline sdks/python/apache_beam/io/gcp/bigquery_test.py]: A better name might be "test_dofn_client_process_performs_batching".

File: sdks/python/apache_beam/io/gcp/bigquery_test.py [hunk 14]
```diff
@@ -824,6 +824,138 @@ def test_get_or_create_dataset_fetched(self):
...
+    fn.process({'month': 1})
+
+    self.assertTrue(client.tables.Get.called)
+    # InsertRows not called as batch size is not hit
+    self.assertFalse(client.tabledata.InsertAll.called)
+
+    fn.finish_bundle()
+    # InsertRows called in finish bundle
+    self.assertTrue(client.tabledata.InsertAll.called)
+
```
Comment (chamikaramj) [inline sdks/python/apache_beam/io/gcp/bigquery_test.py]: Also, add a test that writes zero records.

File: sdks/python/apache_beam/io/gcp/bigquery.py [hunk 15]
```diff
@@ -1134,3 +1139,110 @@ def convert_row_to_dict(self, row, schema):
...
+  def start_bundle(self):
+    self._rows_buffer = []
+    # Transform the table schema into a bigquery.TableSchema instance.
+    if isinstance(self.schema, basestring):
+      # TODO(silviuc): Should add a regex-based validation of the format.
+      table_schema = bigquery.TableSchema()
+      schema_list = [s.strip(' ') for s in self.schema.split(',')]
+      for field_and_type in schema_list:
+        field_name, field_type = field_and_type.split(':')
+        field_schema = bigquery.TableFieldSchema()
```
Comment (chamikaramj) [inline sdks/python/apache_beam/io/gcp/bigquery.py]: Please add tests for schema handling logic here.

File: sdks/python/apache_beam/io/gcp/bigquery.py (hunk 1 above)
Comment (sb2nov) [inline sdks/python/apache_beam/io/gcp/bigquery.py]: Done

File: sdks/python/apache_beam/io/gcp/bigquery.py (hunk 10 above)
Comment (sb2nov) [inline sdks/python/apache_beam/io/gcp/bigquery.py]: This is based on what Java had

File: sdks/python/apache_beam/io/gcp/bigquery.py (hunk 3 above)
Comment (sb2nov) [inline sdks/python/apache_beam/io/gcp/bigquery.py]: Removed

File: sdks/python/apache_beam/io/gcp/bigquery.py (hunk 4 above)
Comment (sb2nov) [inline sdks/python/apache_beam/io/gcp/bigquery.py]: I would have just liked to deprecate this string schema thing and asked everyone to create a table schema object but that is a larger change

File: sdks/python/apache_beam/io/gcp/bigquery.py (hunk 5 above)
Comment (sb2nov) [inline sdks/python/apache_beam/io/gcp/bigquery.py]: not in the string schema input

File: sdks/python/apache_beam/io/gcp/bigquery.py (hunk 6 above)
Comment (sb2nov) [inline sdks/python/apache_beam/io/gcp/bigquery.py]: removed

File: sdks/python/apache_beam/io/gcp/bigquery.py (hunk 7 above)
Comment (sb2nov) [inline sdks/python/apache_beam/io/gcp/bigquery.py]: yes it does. I was running into issues due to write truncate being a delate and recreate

File: sdks/python/apache_beam/io/gcp/bigquery.py (hunk 8 above)
Comment (sb2nov) [inline sdks/python/apache_beam/io/gcp/bigquery.py]: changed to debug

File: sdks/python/apache_beam/io/gcp/bigquery.py (hunk 11 above)
Comment (sb2nov) [inline sdks/python/apache_beam/io/gcp/bigquery.py]: Done

File: sdks/python/apache_beam/io/gcp/bigquery_test.py (hunk 13 above)
Comment (sb2nov) [inline sdks/python/apache_beam/io/gcp/bigquery_test.py]: Done

File: sdks/python/apache_beam/io/gcp/bigquery_test.py (hunk 12 above)
Comment (sb2nov) [inline sdks/python/apache_beam/io/gcp/bigquery_test.py]: Done

File: sdks/python/apache_beam/io/gcp/bigquery_test.py (hunk 14 above)
Comment (sb2nov) [inline sdks/python/apache_beam/io/gcp/bigquery_test.py]: Done

File: sdks/python/apache_beam/io/gcp/bigquery.py (hunk 2 above)
Comment (sb2nov) [inline sdks/python/apache_beam/io/gcp/bigquery.py]: Done

File: sdks/python/apache_beam/io/gcp/bigquery.py (hunk 4 above)
Comment (sb2nov) [inline sdks/python/apache_beam/io/gcp/bigquery.py]: Done

File: sdks/python/apache_beam/io/gcp/bigquery.py (hunk 15 above)
Comment (sb2nov) [inline sdks/python/apache_beam/io/gcp/bigquery.py]: Done

File: sdks/python/apache_beam/io/gcp/bigquery.py (hunk 9 above)
Comment (sb2nov) [inline sdks/python/apache_beam/io/gcp/bigquery.py]: Done

**Comment by chamikaramj** [comment]
LGTM. Thanks.
